nautobot_golden_config/utilities/helper.py

```diff
@@ -139,3 +150,103 @@ def get_device_to_settings_map(queryset):
+    """
+    try:
+        module_name, model_name = obj_type.split(".")
+    except ValueError:  # not enough values to unpack (expected 2, got 1). We get this error if there is no "." in the obj_type
+        return f"Incorrect object type format {obj_type}. The correct format is 'module_name.model_name'."
+    try:
+        model = apps.get_model(module_name, model_name)
+    except ObjectDoesNotExist:
+        return f"Model {model_name} does not exist."
+    # get the object behind the model using the pk
+    try:
+        obj = model.objects.get(id=obj_id)
+    except ObjectDoesNotExist:
+        return f"Object wit ID {obj_id} does not exist."
+
+    # Get user permissions, terminate early if they do not have permission to view
+    app_label = obj._meta.app_label
+    model_name = obj._meta.model_name
+    permission_required = f"{app_label}.view_{model_name}"
+    # Bypass restriction for superusers and exempt views
+    if user.is_superuser or permission_is_exempt(permission_required):
+        pass
+    # User is anonymous or has not been granted the requisite permission
+    elif not user.is_authenticated or permission_required not in user.get_all_permissions():
+        return "You have no permission to read secrets. This incident will be reported."
+
+    # look through the attributes of the model to find if it has an id relationship
+    secrets_group = None
+    secrets_group_id = obj.__dict__.get("secrets_group_id")
+    if secrets_group_id:
+        secrets_group = SecretsGroup.objects.get(id=secrets_group_id)
+
+    # if the secret is not in the attributes, it may be in the relationships
+    for value in obj.get_relationships_data()["destination"].values():
+        # Find the relationship to secrets group
+        if isinstance(value["value"], SecretsGroup):
+            secrets_group = SecretsGroup.objects.get(slug=value["value"])
+            break
+
+    if secrets_group:
+        return secrets_group.get_secret_value(
+            access_type=secret_access_type,
+            secret_type=secret_type,
+            obj=obj,
+        )
+
+    return None
+
+
+def render_secrets(config_to_render: str, user: User):
+    """Renders secrets using the get_secrets filter, and Netutils ones."""
+    jinja_env = jinja2.Environment(autoescape=True)
+
+    # Wrapper for get_secret filter that includes user argument to ensure
+    # that secrets are only rendered by authorized users.
+    user_gets_secret = partial(get_secret, user)
+
+    for name, func in jinja2_convenience_function().items():
+        jinja_env.filters[name] = func
+
+    jinja_env.filters["get_secret"] = user_gets_secret
+    template = jinja_env.from_string(config_to_render)
+    return template.render()
+
+
+def get_config_to_push(config_to_render: str, user: User, device: Device):
+    """Renders final configuration push artifact from intended configuration."""
+    config_to_push = render_secrets(config_to_render, user)
```

Comment:
This to me should not be a default. There should imho, be a configuration that "subscribes" to each change

Reply:
how would you signal it? 
This rendering step is really light, and if you don't have the `get_secret` filter nothing happens. I think that adding a config option for "render_secrets" would overcomplicate it, but I'm open to it if it betters align with the project.